src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -330,14 +301,41 @@ private HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Uri? prox
             }
         }
 
+        // Picks the value of the 'server.address' tag following rules specified in
+        // https://github.com/open-telemetry/semantic-conventions/blob/728e5d1/docs/http/http-spans.md#http-client-span
+        // When there is no proxy, we need to prioritize the contents of the Host header.
+        private static string? GetTelemetryServerAddress(HttpRequestMessage request, HttpConnectionKey key)
+        {
+            if (GlobalHttpSettings.MetricsHandler.IsGloballyEnabled || GlobalHttpSettings.DiagnosticsHandler.EnableActivityPropagation)
+            {
+                Uri? uri = request.RequestUri;
+                Debug.Assert(uri is not null);
+
+                if (key.ProxyUri is not null)
+                {
+                    return uri.IdnHost;
+                }
```

Comment:
This doesn't look correct for the HTTP request over an HTTP proxy case.

https://github.com/antonfirsov/runtime/blob/0b23873530b1d0c51122379292e093832f0bc339/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L282-L285

We'll report the server address as the first request we use that connection for, which might be misleading since it's actually a shared connection.

Reply:
I will then revert this case to the original behavior [which reports the proxy host](https://github.com/dotnet/runtime/blob/23323e91728ba356ea8ed029631a63a55d90907a/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.cs#L560):

```suggestion
                if (key.ProxyUri is not null)
                {
                    return key.ProxyUri.IdnHost;
                }
```

> a shared connection

Note that the [current spec](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/http-metrics.md#metric-httpclientconnectionduration) wants us to meet an impossible expectation here

> `server.address` SHOULD represent the server address behind any intermediaries

Raised a point under https://github.com/open-telemetry/semantic-conventions/issues/2443#issuecomment-3083990510.

Reply:
Thank you for following up there.
If we want to still prefer the host, we could also flip the order and use `SslHostName` even if there's a proxy, maybe that's closer to the intension?